From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: fgei <fgei@gmail.com>
Date: Fri, 28 Jan 2022 20:52:56 +0000
Subject: [PATCH] Implement UI for JIT site settings

Adapted from GrapheneOS Vanadium patch 0113-Implement-UI-for-JIT-site-settings.patch for the Vandium browser port.

---
 .../res/xml/site_settings_preferences.xml     |  4 ++++
 .../ContentSettingsResources.java             | 13 ++++++++++
 .../site_settings/SingleCategorySettings.java |  5 ++++
 .../site_settings/SingleWebsiteSettings.java  | 24 +++++++++++++++++++
 .../site_settings/SiteSettingsCategory.java   |  8 ++++++-
 .../site_settings/SiteSettingsUtil.java       |  1 +
 .../browser_ui/site_settings/Website.java     |  6 +++++
 .../android/website_preference_bridge.cc      |  1 +
 .../strings/android/site_settings.grdp        | 24 +++++++++++++++++++
 .../core/browser/content_settings_registry.cc |  2 +-
 .../android/page_info_controller_android.cc   |  3 +++
 components/page_info/page_info.cc             |  6 +++++
 components/page_info/page_info_ui.cc          |  2 ++
 components/site_settings_strings.grdp         |  6 +++++
 14 files changed, 103 insertions(+), 2 deletions(-)

diff --git a/components/content_settings/core/browser/content_settings_registry.cc b/components/content_settings/core/browser/content_settings_registry.cc
index 67f34ff0793cf..c8edc7934e421 100644
--- a/components/content_settings/core/browser/content_settings_registry.cc
+++ b/components/content_settings/core/browser/content_settings_registry.cc
@@ -610,7 +610,7 @@ void ContentSettingsRegistry::Init() {
            PermissionSettingsInfo::EXCEPTIONS_ON_SECURE_ORIGINS_ONLY);
 
   Register(ContentSettingsType::JAVASCRIPT_JIT, "javascript-jit",
-           CONTENT_SETTING_ALLOW, WebsiteSettingsInfo::UNSYNCABLE,
+           CONTENT_SETTING_BLOCK, WebsiteSettingsInfo::UNSYNCABLE,
            /*allowlisted_primary_schemes=*/{},
            /*valid_settings=*/{CONTENT_SETTING_ALLOW, CONTENT_SETTING_BLOCK},
            WebsiteSettingsInfo::TOP_ORIGIN_ONLY_SCOPE,
diff --git a/components/page_info/page_info.cc b/components/page_info/page_info.cc
index c2c5b9d3039c1..5ee8d230ab32c 100644
--- a/components/page_info/page_info.cc
+++ b/components/page_info/page_info.cc
@@ -109,6 +109,7 @@ ContentSettingsType kPermissionType[] = {
     ContentSettingsType::SENSORS,
     ContentSettingsType::NOTIFICATIONS,
     ContentSettingsType::JAVASCRIPT,
+    ContentSettingsType::JAVASCRIPT_JIT,
 #if !BUILDFLAG(IS_ANDROID)
     ContentSettingsType::IMAGES,
 #endif
@@ -1359,6 +1360,11 @@ bool PageInfo::ShouldShowPermission(
   }
 #endif  // BUILDFLAG(IS_CHROMEOS)
 
+  // Always show JIT settings UI when when it has a site-specific override.
+  if (info.type == ContentSettingsType::JAVASCRIPT_JIT) {
+    return true;
+  }
+
   const bool is_incognito =
       web_contents_->GetBrowserContext()->IsOffTheRecord();
 #if BUILDFLAG(IS_ANDROID)
diff --git a/components/page_info/page_info_ui.cc b/components/page_info/page_info_ui.cc
index 213cafae77f4a..ad2633992abcb 100644
--- a/components/page_info/page_info_ui.cc
+++ b/components/page_info/page_info_ui.cc
@@ -139,6 +139,8 @@ base::span<const PageInfoUI::PermissionUIInfo> GetContentSettingsUIInfo() {
        IDS_SITE_SETTINGS_TYPE_COOKIES_MID_SENTENCE},
       {ContentSettingsType::JAVASCRIPT, IDS_SITE_SETTINGS_TYPE_JAVASCRIPT,
        IDS_SITE_SETTINGS_TYPE_JAVASCRIPT_MID_SENTENCE},
+      {ContentSettingsType::JAVASCRIPT_JIT, IDS_SITE_SETTINGS_TYPE_JAVASCRIPT_JIT,
+       IDS_SITE_SETTINGS_TYPE_JAVASCRIPT_JIT_MID_SENTENCE},
       {ContentSettingsType::POPUPS, IDS_SITE_SETTINGS_TYPE_POPUPS_REDIRECTS,
        IDS_SITE_SETTINGS_TYPE_POPUPS_REDIRECTS_MID_SENTENCE},
       {ContentSettingsType::GEOLOCATION, IDS_SITE_SETTINGS_TYPE_LOCATION,
diff --git a/components/site_settings_strings.grdp b/components/site_settings_strings.grdp
index 4261804b4e62a..26941202af2c0 100644
--- a/components/site_settings_strings.grdp
+++ b/components/site_settings_strings.grdp
@@ -121,6 +121,12 @@
   <message name="IDS_SITE_SETTINGS_TYPE_JAVASCRIPT_MID_SENTENCE" desc="The label used for JavaScript site settings controls when used mid-sentence.">
     javascript
   </message>
+  <message name="IDS_SITE_SETTINGS_TYPE_JAVASCRIPT_JIT" desc="The label used for JavaScript JIT site settings controls.">
+    JavaScript JIT
+  </message>
+  <message name="IDS_SITE_SETTINGS_TYPE_JAVASCRIPT_JIT_MID_SENTENCE" desc="The label used for JavaScript JIT site settings controls when used mid-sentence.">
+    javascript JIT
+  </message>
   <message name="IDS_SITE_SETTINGS_TYPE_JAVASCRIPT_OPTIMIZER" desc="The label for the JavaScript optimizer site settings controls.">
     JavaScript optimization &amp; security
   </message>
